logutils/namespaces_tree.go

```diff
@@ -0,0 +1,137 @@
+package logutils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/zapcore"
+)
+
+var namespacesRegex = regexp.MustCompile(`^([a-zA-Z0-9_\.]+:(debug|info|warn|error))(,[a-zA-Z0-9_\.]+:(debug|info|warn|error))*$`)
+var errInvalidNamespacesFormat = fmt.Errorf("invalid namespaces format")
+
+type namespacesTree struct {
+	namespaces map[string]*namespaceNode
+	minLevel   zapcore.Level
+	mu         sync.RWMutex
+}
+
+type namespaceNode struct {
+	level    zapcore.Level
+	children map[string]*namespaceNode
+}
+
+func newNamespacesTree() *namespacesTree {
+	return &namespacesTree{
+		namespaces: make(map[string]*namespaceNode),
+		minLevel:   zapcore.InvalidLevel,
+	}
+}
+
+// buildNamespacesTree creates a new tree instance from a string of namespaces.
+// The namespaces string should be a comma-separated list of namespace:level pairs,
+// where level is one of: "debug", "info", "warn", "error".
+//
+// Example: "namespace1:debug,namespace2.namespace3:error"
+func buildNamespacesTree(namespaces string) (*namespacesTree, error) {
+	tree := newNamespacesTree()
+
+	err := tree.Rebuild(namespaces)
+	if err != nil {
+		return nil, err
+	}
+
+	return tree, nil
+}
+
+// LevelFor finds the log level for the given namespace by searching for the closest match in the tree.
+// In case of no match, it returns zapcore.InvalidLevel.
+func (t *namespacesTree) LevelFor(namespace string) zapcore.Level {
```

Comment:
This function sounds not quite simple and fast, there's loop, string operations and node lookups. And this will be executed on each log operation.
Do you think we should provide some cache for cache for `namespace`->`logLevel` here?

Reply:
@igor-sirotin, I've just added a benchmark for a pretty complex filter (12 entries in the `namespacesTree`), results:
```
BenchmarkNamespacesFilteringCore-16        41726             30076 ns/op            2238 B/op         35 allocs/op
BenchmarkNamespacesZapCore-16              39079             29178 ns/op            2019 B/op         29 allocs/op
```

`namespacesFilteringCore` is just a bit slower than vanilla `zap.Core`.

Reply:
Alright, sounds not that bad. And good job with benchmarking things 👍

Reply:
:pray: